src/libsyntax/parse/token.rs

```diff
@@ -273,17 +310,32 @@ impl Token {
         }
     }
 
-    pub fn ident(&self) -> Option<ast::Ident> {
+    fn ident_common(&self, allow_raw: bool) -> Option<ast::Ident> {
```

Comment:
This should be just `pub fn ident(&self) -> Option<(ast::Ident, /* is_raw */ bool)>` IMO.
It will avoid introducing all those many little helper functions.
(The `fn nonraw_ident` vs `fn ident` distinction also looks error-prone.)